File: 1-working-with-list/p06_is_palindrome.ml

let rec rev l =
  match l with
  | [] -> []
  | hd :: tl -> (rev tl) @ [hd];;

let is_palindrome lst =
  (rev lst) = lst;;

let is_palindrome' lst =
  let rev l =
    let rec aux acc = function
      | [] -> acc
      | hd::tl -> aux (hd::acc) tl
    in
    aux [] l
  in
  rev lst = lst;;

is_palindrome [ "x" ; "a" ; "m" ; "a" ; "x" ];;
not (is_palindrome [ "a" ; "b" ]);;

is_palindrome' [ "x" ; "a" ; "m" ; "a" ; "x" ];;
not (is_palindrome' [ "a" ; "b" ]);;

File: 1-working-with-list/p06_is_palindrome.scala

object Main {

  def main(args: Array[String]) {
    println(isPalindrome(List("x", "a", "m", "a", "x")))
    println(isPalindrome(List("a", "b")))
  }

  def rev[T](l: List[T]): List[T] = {
    l match {
      case Nil => Nil
      case hd :: rest => rev(rest) :+ hd
    }
  }

  def isPalindrome[T](l: List[T]): Boolean = {
    rev(l) == l
  }
}
